google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,6 +182,274 @@ class ResumableAsyncStreamingReadWriteRpc {
+                std::unique_lock<std::mutex> lk{mu_};
+                CompleteUnsatisfiedOpsLockHeld(Status(), lk);
+              })));
+          break;
+        case State::kInitialized:
+          stream_state_ = State::kShutdown;
+          if (in_progress_read_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_read_->get_future()));
+          }
+          if (in_progress_write_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_write_->get_future()));
+          }
+          std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+              stream = std::move(stream_);
+          CompleteUnsatisfiedOpsLockHeld(Status(), lk);
+          root_future = root_future.then([stream](future<void>) {
+            return future<void>(stream->Finish());
+          });
+      }
+    }
+    return root_future;
+  }
+
+  void SetReadWriteFuturesLockHeld(std::unique_lock<std::mutex>& lk) {
+    absl::optional<promise<void>> read_reinit_done;
+    absl::optional<promise<void>> write_reinit_done;
+
+    {
+      read_reinit_done.swap(read_reinit_done_);
+      write_reinit_done.swap(write_reinit_done_);
+    }
+    lk.unlock();
+    if (read_reinit_done) read_reinit_done->set_value();
+
+    if (write_reinit_done) write_reinit_done->set_value();
+    lk.lock();
+  }
+
+  void CompleteUnsatisfiedOpsLockHeld(Status status,
+                                      std::unique_lock<std::mutex>& lk) {
+    SetReadWriteFuturesLockHeld(lk);
+    status_promise_.set_value(std::move(status));
+  }
+
+  void FinishRetryPromiseLockHeld(std::unique_lock<std::mutex>& lk) {
+    assert(retry_promise_.has_value());
+    promise<void> retry_promise = std::move(*retry_promise_);
+    retry_promise_.reset();
+    lk.unlock();
+    retry_promise.set_value();
+    lk.lock();
+  }
+
+  void AttemptRetry(Status const& status,
+                    std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    {
+      std::unique_lock<std::mutex> lk{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return FinishRetryPromiseLockHeld(lk);
+      }
+      assert(stream_state_ == State::kRetrying);
+    }
+
+    if (!retry_policy->IsExhausted() && retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void>) {
+            Initialize(retry_policy, backoff_policy);
+          });
+      return;
+    }
+
+    std::unique_lock<std::mutex> lk{mu_};
+    FinishRetryPromiseLockHeld(lk);
+    if (stream_state_ == State::kShutdown) return;
+    stream_state_ = State::kShutdown;
+
+    CompleteUnsatisfiedOpsLockHeld(status, lk);
+  }
+
+  void OnInitialize(
+      StatusOr<std::shared_ptr<UnderlyingStream>> start_initialize_response,
+      std::shared_ptr<UnderlyingStream> stream,
```

Comment:
remove this argument, you don't need to call Finish(), that's the initializer's responsibility